OpenMP 5 target offload backend

jkelling opened the pull request:
This PR turns the existing OpenMP4 backend into one with target offload and renames it.

## AccCpuOmp4 -> AccOmp5

* Practically, most of this might also work with OpenMP 4.0, depending on how the compiler interprets the standard. Also there is currently no compiler which actually supports OpenMP 5.0 (Clang 11 claims to, but the support is incomplete). Some compilers claiming OpenMP 4.5/4.0 may work partially, or we can push their interpretation of the standard into a direction that works for us. This is why the `_OPENMP` guards are unchanged (still OpenMP 4.0). Technically, C++ TMP requires OpenMP 5.0 to undisputably work and there are some other problems in the earlier versions, which is why the backend is named Omp5.
* Since this can run on any device, I removed the `Cpu`(`Gpu`) part alltogether and replaced this bit by `ANY` in the cmake option.

## CMake

* Adds option (string) `ALPAKA_OFFLOAD_MAX_BLOCK_SIZE` which set the number blocks per thread that the backend reports as maximum.
  * If the kernels are invoked with more threads per block than the runtime will actually run, some work will not get done (=> wrong results). There is a check in place from the old Omp4 backend in ALPAKA_DEBUG mode printing a warning. Some runtimes will also warn about this.
* Adds option (bool) `ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST` which tells the code, that the target it is compuled for is the host, thus allowing things like asserts in target code. There is no OpenMP standard way for the code to get information about the target at compile that that I am aware of.

## CI

* Because compiler support is very limited, only this backend can only be tested with clang 10 in the CI. I changed the `linux_clang-10_release` job to include it and added `-omp5` to the name to mark it.
* Had to change the clang install script to pull in the requirements for target offload for versions >= 10.
  * Because some clang versions are given as e.g. "4.0" in ci.yml, I had to use this condition to check the version:
    https://github.com/jkelling/alpaka/blob/4c2de74a549b3e998b0cfb0bf8eea860d7bdc298/script/install_clang.sh#L21
      ```bash
     [ "$( echo "${ALPAKA_CI_CLANG_VER} < 10" | bc )" == 1 ]
     ```
    Should one always assume, that version numbers can be float when comparing, or should we rather, or can we even, change version number to be only major version ints?
* `ALPAKA_OFFLOAD_MAX_BLOCK_SIZE=1` because the runtime in the CI will not run more than one thread per team even though `OMP_NUM_THREADS=4` and it is running multiple teams.

## Backend Properties

* `AccCpuOmp4` -> `AccOmp5`
* Adds `DevOmp5`, to handle OMP device ids. Also `PltfOmp5`
* Adds `BufOmp5` for OMP target allocations.
* Currently uses `QueueGenericThreads*<DevOmp5>`. An asynchronous queue could later be implemented based on OMP tasking constructs.

## Tests

* Turns `cpu::detail::EventHostManualTriggerCpuImpl` into a template with respect to the device: `DevCpu` or `DevOmp5`.

jkelling committed:
TaskKernelCpuOmp4: Replace kernel std::bind:s

not save/allowed in target region

jkelling committed:
OMP4 target: Add terget memory (WIP)

jkelling committed:
AccCpuOmp4: Implement Omp4 Traits for non-host offload

jkelling committed:
Omp4: Add WorkDivOmp4BuildIn

not sure if it compiles, triggers an internal compiler error in AOMP

jkelling committed:
omp4/Copy: Add fix TDimm = 1 (remove)

jkelling committed:
PltfOmp4: Add activation guard

jkelling committed:
QueueOmp4Blocking::ctor: Add noexcept

jkelling committed:
buf/omp4/Copy: Fix copy host deviceID

jkelling committed:
Add README_OMP5.md

jkelling committed:
Update README_OMP5 instructions

jkelling committed:
README_OMP5: split status col

jkelling committed:
README_OMP5: Add vectorAdd output explanation

jkelling committed:
WorkDivOmp4BuiltIn: store cpy of m_threadElemExtent, not &

jkelling committed:
Omp4: Fix adjust threadElementCount if numteams is limited

jkelling committed:
omp4:TaskKernel Add copy and map m_args

jkelling committed:
omp4::TaskKernel: Add assert(enough OMP theads), do not adjust

jkelling committed:
omp4::TaskKernel: Add loop over teams to stretch blockcount

jkelling committed:
README_OMP5: Update compiler state

jkelling committed:
README_OMP5: Add OMP5 compiler flags

jkelling committed:
README_OMP5.md: Add compiler status on claix

jkelling committed:
omp4::TaskKernel: Fix args map direction

update README_OMP5 (claix)

jkelling committed:
README_OMP5: Update LLVM9(claix) nvptx result for -O2

jkelling committed:
Omp4: Fixes for test compilation

Add Nd grid/block/thread, buf

jkelling committed:
Omp4: Add nD Copy using omp_target_memcpy_rect (no pitch)

jkelling committed:
omp4: Disable separableCompilation test for OMP4

jkelling committed:
omp4: Add buf/mem/Set.hpp

jkelling committed:
omp4: Fix atomic Hierarchy

set grid level to AtomicOmpBuiltIn

jkelling committed:
omp4: Fixes

jkelling committed:
omp4: Fix mem/buf/Set

jkelling committed:
omp4: mem/buf/Copy: Add use of getExtentVec (disabled du to missing cast)

jkelling committed:
omp4: Fix type matches

jkelling committed:
omp4: disable AtomicTest, because of LLVM NVPTX bug:

https://bugs.llvm.org/show_bug.cgi?id=44219

jkelling committed:
omp4: disable test KernelStdFunction

no function pointers allowed in target region, no abort()

jkelling committed:
omp4: Add CreateStaticDevMemView<dev::DevOmp4>

jkelling committed:
omp4: test/unit/rand: disable STL rand

jkelling committed:
omp4: Fix sign conversion warnings

jkelling committed:
AccCpuOmp4: Add dyn smem

jkelling committed:
TaskKernelCpuOmp4: Fix workdiv init of Acc

jkelling committed:
BufOmp4: Fix Nd copy (VecFrom...)

jkelling committed:
IdxGbOmp4BuiltIn: Fix GetIdx<Grid,Blocks> workdiv access

jkelling committed:
omp4/Set: Fix

omp4/Copy: Add nop if size is zero

jkelling committed:
omp4 Buf+Copy: Fix view WIP

jkelling committed:
README_OMP5.md: Add tests; update

jkelling committed:
omp4: Fix compile after rebase

jkelling committed:
omp4::Idx&Dev: Add concepts

jkelling committed:
README_OMP5.md: Update summit

jkelling committed:
AccCpuOmp4::m_gridBlockIdx: Rm mutable (not required)

jkelling committed:
README_OMP5: Add triggering lines for Test error 3.

jkelling committed:
BufOmp4: Fix ALPAKA_DEBUG print

jkelling committed:
omp4::Copy: Fix ALPAKA_FULLDEBUG

jkelling committed:
TaskKernelCpuOmp4: Add compilte-time check against IBM XL

.. num_threads does not work

jkelling committed:
Update README_OMP5 about IBM XL on summit

jkelling committed:
README_OMP5: Fix XL targets, Add ppc8le

jkelling committed:
TaskKernelCpuOmp4: Add static_cast for arg of ompSetNumThreads

jkelling committed:
README_OMP5: Add gcc 9.2 Vec map error

jkelling committed:
README_OMP5: Add numbers to section

jkelling committed:
PltfOmp4: Add concept

jkelling committed:
DevOmp4: implement QueueType trait

jkelling committed:
Add guard ALPAKA_ACC_CPU_BT_OMP4_ENABLED around omp4 filters in tests

jkelling committed:
README_OMP5: do not disable acc SEQ_SEQ

jkelling committed:
BlockSharedMemStOmp4::alloc: Fix race condition...

.. for alloc() calls in short sequence.

jkelling committed:
omp4/Copy: Move impl into Task, Rm Equeue

jkelling committed:
omp4: Fixes

jkelling committed:
BlockSharedMemDynOmp4: Reduce static smem alloc to 30kB

jkelling committed:
AccCpuOmp4: Rm include StdLibLock

jkelling committed:
README_OMP5: Update AOMP 0.7-5 state

jkelling committed:
TaskKernelCpuOmp4: Add cpy m_kernelFnObj to local scope

.. for mapping
Rm explicit map
Add explicit capurte kernelFnObj, acc in lambda

jkelling committed:
README_OMP5: Add llvm fix: use c++14

jkelling committed:
TaskKernelCpuOmp4: ensure numTeams>0

no limit to blockThreadCount

jkelling committed:
BufOmp4: alloc'd message to std::cout

jkelling committed:
OMP4: Add target device selection

jkelling committed:
README_OMP5: Add seq_seq enable to cmake flags

jkelling committed:
AccCpuOmp4: Add param ALPAKA_OFFLOAD_MAX_BLOCK_SIZE

jkelling committed:
DevOmp4,PltfOmp4: Fix  iDevice, Add num_devices==0 handling

jkelling committed:
EventHostManualTrigger.hpp: Add guard for OMP4

jkelling committed:
omp4::Acc: reverse workdiv order

jkelling committed:
Fix TaskKernelCpuOmp4 traits Dev,Pltf

jkelling committed:
PltfOmp4: Fix GetDevByIdx(devIdx) cast warning

jkelling committed:
AccCpuOmp4: Use WorkDivMembers

jkelling committed:
omp4/Copy: Add static_cast<size_t>(m_extendWidthbytes)

jkelling committed:
omp4: Add QueueOmp4*Blocking

use QueueGenericThreads* for omp4 queues

jkelling committed:
Add alias EventOmp4

jkelling committed:
Add CurrentThreadWaitFor for DevOmp4

jkelling committed:
omp4: Fix GetDevCount

have to fake 1 device if omp_get_num_devices() == 0

jkelling committed:
Rename AccCpuOmp4 -> AccOmp5 (and related)

jkelling committed:
Rm superfluos BOOST_NO_CXX14_RETURN_TYPE_DEDUCTION checks

jkelling committed:
README_OMP5: Update OMP5

jkelling committed:
Add cmake variable ALPAKA_OFFLOAD_MAX_BLOCK_SIZE

jkelling committed:
Add GetWarpSize<DevOmp5> stub

jkelling committed:
Omp5: Use BlockSharedMemDynMember

jkelling committed:
GetAccDevProps<AccOmp5<>>: Add staticAllocBytes()

jkelling committed:
OMP5: s/ALPAKA_ACC_CPU_BT_OMP4_ENABLE/ALPAKA_ACC_ANY_BT_OMP5_ENABLE/

jkelling committed:
Omp5: Add ALPAKA_DEBUG guards to debug prints

jkelling committed:
AtomicTest: Enable for Omp5 if _OFFLOAD_ASSUME_HOST

jkelling committed:
AccOmp5: use IntrinsicFallback

jkelling committed:
CI: Add Omp5 clang10 x86

jkelling committed:
DevOmp5: Add conceptDev

jkelling committed:
TestAccs: Rm diasble OMP5 for CI

jkelling committed:
Disable tests with multiple object files for OMP5+GCC10

... bug with mssing/not-found target symbols

jkelling committed:
Omp5: Disable test atomic in TestAccs.hpp

jkelling committed:
Disable test separableCompilation in CMakeLists.txt

jkelling committed:
ExampleDefaultAcc.hpp: Omp4->Omp5

jkelling committed:
TaskKernelOmp5: use num_threads clause ...

... instead of omp_set_num_threads. keep XL workaround in place

jkelling committed:
TaskKernelOmp5: Add check num_[teams,threads] in CI

jkelling committed:
Add IdxGbLinear; cleanup AccOmp5 Gb Idx handling

TaskKernelOmp5: Rm outer teams loop

jkelling commented:
The CI job "linux_clang-10_release" is not going to happen, because I renamed it to "linux_clang-10_release-omp5", why is this listed as required?

BenjaminW3 commented:
The selection which jobs are required is part of repository settings. They are mentioned there by name.
This is somehow fragile but it is what github offers at the moment. When we want to rename a job, we have to change the settings as well. I can do this, once this PR is ready or has been merged.

BenjaminW3 commented:
Clang changed the scheme of their ubuntu packet names from two numbers seperated by a dot (`clang-4.0`) to single number (`clang-9`). That is the reason for the different numbers in the yml file. `4.0` without quotes seems to be interpreted as floating point number and forwarded as `4` which does not match the packet names anymore.

BenjaminW3 commented:
Could we use this PR as a general discussion board but split off some uncritical changes into independent PRs that can easily be merged?
The first change I would like to get merged independently is the change you did to `EventHostManualTriggerCpuImpl`. I may find more things later.

BenjaminW3 commented on thirdParty/catch2/include/catch2/catch.hpp:
```diff
@@ -11392,6 +11393,8 @@ FP step(FP start, FP direction, uint64_t steps) {
 bool marginComparison(double lhs, double rhs, double margin) {
     return (lhs + margin >= rhs) && (rhs + margin >= lhs);
+    // auto ulpDiff = std::abs((FP)(lc.i - rc.i));
+    // return ulpDiff <= maxUlpDiff;
```
Could you pleas remove those comments?

jkelling replied:
Removed.

BenjaminW3 commented on thirdParty/catch2/include/catch2/catch.hpp:
```diff
@@ -11469,14 +11472,14 @@ namespace Floating {
-            write(ret, step(static_cast<float>(m_target), static_cast<float>( INFINITY), m_ulps));
+            write(ret, step(static_cast<float>(m_target), -std::numeric_limits<float>::infinity(), m_ulps));
+            ret << ", ";                                                                        
+            write(ret, step(static_cast<float>(m_target),  std::numeric_limits<float>::infinity(), m_ulps));
```
Could you create a PR with those changes to the original catch2 repository?

jkelling replied:
These changes to catch2 are only required for some special cases:
* This one when compiling for offload to nvptx using clang.
* The other one with IMB's XL compiler.
Neither should be an issue in the CI. I will test this theory by pushing a version of this branch without these changes later.

These patches to not apply to the current catch2 master. I will have to open issues rather than a PR.

BenjaminW3 commented on thirdParty/catch2/include/catch2/catch.hpp:
```diff
@@ -1298,7 +1299,7 @@ struct AutoReg : NonCopyable {
                     int index = 0;\
                     using expander = int[];\
-                    (void)expander{(Catch::AutoReg( Catch::makeTestInvoker( &TestName<Types>::test ), CATCH_INTERNAL_LINEINFO, #ClassName, Catch::NameAndTags{ Name " - " + std::string(INTERNAL_CATCH_STRINGIZE(TmplList)) + " - " + std::to_string(index), Tags } ), index++, 0)... };/* NOLINT */ \
+                    (void)expander{0, (Catch::AutoReg( Catch::makeTestInvoker( &TestName<Types>::test ), CATCH_INTERNAL_LINEINFO, #ClassName, Catch::NameAndTags{ Name " - " + std::string(INTERNAL_CATCH_STRINGIZE(TmplList)) + " - " + std::to_string(index), Tags } ), index++, 0)... };/* NOLINT */ \
```
What exactly is this doing?

jkelling replied:
It makes sure the list is never empty. This is a workaround for a bug in IBM XL, which cannot deal with empty initializer lists. See: https://github.com/alpaka-group/alpaka/pull/1126/commits/8da30a1429ef8df4b820c0c3a5e4d784d79f1306

BenjaminW3 commented on thirdParty/catch2/include/catch2/catch.hpp:
```diff
@@ -11342,7 +11343,7 @@ namespace {
         }
 
-        auto ulpDiff = std::abs(lc - rc);
+        auto ulpDiff = std::abs((FP)(lc - rc));
```
This should probably also be added to the catch2 original.

BenjaminW3 commented on test/common/devCompileOptions.cmake:
```diff
@@ -139,6 +139,10 @@ ELSE()
+        # Triggers for all instances of ALPAKA_DEBUG_MINIMAL_LOG_SCOPE and similar macros followed by semicolon
+        IF(CLANG_VERSION_MAJOR GREATER 7)
+            LIST(APPEND ALPAKA_DEV_COMPILE_OPTIONS "-Wno-extra-semi-stmt")
+        ENDIF()
```
Can we add this in a separate PR?

jkelling replied:
https://github.com/alpaka-group/alpaka/pull/1131

BenjaminW3 commented on script/install_clang.sh:
```diff
@@ -18,7 +18,13 @@ source ./script/set.sh
+    travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install clang-${ALPAKA_CI_CLANG_VER}
+else
+    travis_retry sudo apt-get -y --quiet --allow-unauthenticated --no-install-recommends install \
+        clang-tools-${ALPAKA_CI_CLANG_VER} llvm-${ALPAKA_CI_CLANG_VER} libomp-${ALPAKA_CI_CLANG_VER}-dev
```
We are already installing `libomp-dev` below now (this probably was not the case when you created this branch).
We should probably only install one version.

jkelling replied:
Actualy, I missed this bit. Changed the install script to rely less on the clang version, but take into account whether OMP5 is enbabled: https://github.com/alpaka-group/alpaka/pull/1126/commits/912f6db4e27beb5b2c3840857dee96c44af7a493

BenjaminW3 commented on include/alpaka/queue/QueueOmp5NonBlocking.hpp:
```diff
@@ -0,0 +1,43 @@
+#include <functional>
+#include <mutex>
+#include <condition_variable>
+#include <thread>
```
I am pretty sure we can get rid of all (really all) includes except `QueueGenericThreadsNonBlocking.hpp` here and in the blocking version.

jkelling replied:
Also need `DevOmp5.hpp`, removed the rest.

BenjaminW3 commented on include/alpaka/idx/MapIdx.hpp:
```diff
@@ -157,5 +157,138 @@ namespace alpaka
+                ::mapIdxPitch(
+                    idx,
+                    pitch);
+        }
```
What exactly differentiates those `pitch` versions from the `extent` versions?
It might be confusing for users to see those new `pitch` versions.
We should discuss those additions together with some unit tests in a separate PR.

jkelling replied:
They use pitches in order to work with pitched views. They are used by a memset kernel which is also part of this PR. I created a single PR for both, because the belong together: https://github.com/alpaka-group/alpaka/pull/1132

BenjaminW3 commented on test/common/include/alpaka/test/acc/TestAccs.hpp:
```diff
@@ -108,16 +108,29 @@ namespace alpaka
-#if defined(ALPAKA_ACC_CPU_BT_OMP4_ENABLED) && !defined(ALPAKA_CUDA_CI)
+#if defined(ALPAKA_ACC_ANY_BT_OMP5_ENABLED) && !defined(TEST_UNIT_KERNEL_KERNEL_STD_FUNCTION) \
+        && !( BOOST_COMP_GNUC && ( \
+                /* tests excluded because of GCC10 Oacc / Omp5 target symbol bug with multiple units */ \
```
What exactly is the issue here?

jkelling replied:
For example, the intrinsic test produces the following linker errors (both with OpenMP5 target and OpenACC, the example below is OpenACC):
```
`_ZNK6alpaka6kernel14TaskKernelOaccISt17integral_constantImLm1EElNS_3mem4view12MemSetKernelEJRKhPhRNS_3vec3VecIS3_lEESD_EEclERKNS_3dev7DevOaccE._omp_fn.0' referenced in section `.gnu.offload_funcs' of CMakeFiles/intrinsic.dir/src/Popcount.cpp.o: defined in discarded section `.text._ZNK6alpaka6kernel14TaskKernelOaccISt17integral_constantImLm1EElNS_3mem4view12MemSetKernelEJRKhPhRNS_3vec3VecIS3_lEESD_EEclERKNS_3dev7DevOaccE._omp_fn.0[_ZNK6alpaka6kernel14TaskKernelOaccISt17integral_constantImLm1EElNS_3mem4view12MemSetKernelEJRKhPhRNS_3vec3VecIS3_lEESD_EEclERKNS_3dev7DevOaccE]' of CMakeFiles/intrinsic.dir/src/Popcount.cpp.o
```
There are two compilarion units in this order:
1. Ffs.cpp.o
2. Popcount.cpp.o

MemSetKernel is used in both, the symbols are missing for the second unit. They are always missing for all but the first unit.

I just tested this with a fresh gcc master build. There it appears fixed, so we can expect this to be fixed in GCC 11. Adding another check... .

jkelling replied:
New check: https://github.com/alpaka-group/alpaka/pull/1126/commits/d20db3eeb5bc541aa7e234bb3ea273e2f0e2d489

BenjaminW3 replied:
Do we need this whole test disabling stuff in this PR at all? We are not enabling any gcc builds so we can simply wait for gcc-11 and only enable it there.

jkelling replied:
The same stuff is in the OpenACC PR, where it is needed, because there is currently no other option than gcc10 in the CI for this one. So, it is mostly there for symmetry, although one could try to enable this with gcc10 in the CI, too. The reason I did not is because I can only compile but not run on my local machine. This is because it only compiles with `-foffload=disable` (i.e. no GPU) which will make it generate code for the host only but at runtime it sees a GPU and allocates the memory there, making the tests segfault. In the CI there is no GPU, so it will either work, because it allocates on the host or also fail, because it fails to allocate anything on target. I did not test this, but wanted to keep the option open.

On the other hand: could we make a compile-only Ci job?

BenjaminW3 replied:
Yes, currently we have the variable `ALPAKA_CI_ANALYSIS` which selects whether the alpaka tests should be run or if the self-contained-header build (and some other small things) should be done. This was necessary in Travis CI because both things combined took too long. This may not be necessary now with github actions but this is still available.

BenjaminW3 replied:
I just took a look at the gcc release dates and it lloks like gcc-11 will not be released before May 2021. So waiting for it does not seem to be an option.

jkelling replied:
I added an `ALPAKA_CI_ANALYSIS` job with gcc 10 which includes OMP5.

BenjaminW3 commented on test/unit/rand/src/RandTest.cpp:
```diff
@@ -90,6 +90,9 @@ class RandTestKernel
   !defined(ALPAKA_ACC_GPU_HIP_ENABLED)
+#ifndef ALPAKA_ACC_ANY_BT_OMP5_ENABLED
+        // TODO: These ifdefs are wrong: They will reduce the test to the
+        // smallest common denominator from all enabled backends
```
This is what I would expect. If we configured alpaka with mulitple backends, only the tests which are supported by all of them can run.

jkelling replied:
On the level of whole tests, they are only disabled for single backends, i.e. the test will still be done for the backends which support it even of non-supporting backend are enabled. This did not seem viable at the level of granularity, but I am not sure, hence the discussion.

BenjaminW3 commented on cmake/alpakaCommon.cmake:
```diff
@@ -234,24 +235,27 @@ endif()
         message(FATAL_ERROR "Optional alpaka dependency OpenMP could not be found!")
+        set(ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE OFF CACHE BOOL "Enable the OpenMP 2.0 CPU grid block back-end" FORCE)
+        set(ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE OFF CACHE BOOL "Enable the OpenMP 2.0 CPU block thread back-end" FORCE)
+        set(ALPAKA_ACC_ANY_BT_OMP5_ENABLE OFF CACHE BOOL "Enable the OpenMP 5.0 CPU block and thread back-end" FORCE)
```
Please remove. This is not necessary anymore as we error now above.

jkelling replied:
Removed.

BenjaminW3 commented on cmake/alpakaCommon.cmake:
```diff
@@ -234,24 +235,27 @@ endif()
-                message(FATAL_ERROR "OpenMP 4.0 is required but not supported!")
+         if(ALPAKA_ACC_ANY_BT_OMP5_ENABLED)
+           if(OpenMP_CXX_VERSION VERSION_LESS 4.0)
+               set(ALPAKA_ACC_ANY_BT_OMP5_ENABLE OFF CACHE BOOL "Enable the OpenMP 5.0 CPU block and thread back-end" FORCE)
```
We now error when a requested backend is not available. Please adapt this line here.

jkelling replied:
> The selection which jobs are required is part of repository settings. They are mentioned there by name.
> This is somehow fragile but it is what github offers at the moment. When we want to rename a job, we have to change the settings as well. I can do this, once this PR is ready or has been merged.

Ok. My rationale about changing this name is, that this is the only job with omp5 enabled, and will be one of only a few for a while. If for a future PR clang-10 fails while others succeed, it may be good to see, that it includes the omp5  backend.

jkelling replied:
Done. Also added a warning for OpenMP < 5.0: https://github.com/alpaka-group/alpaka/pull/1126/commits/7fcec9209569693dfd5755a377581eb16663e5e4

BenjaminW3 commented on README_OMP5.md:
```diff
@@ -0,0 +1,190 @@
+# Building configuring for the OMP5 backend
+
+To make the build system enable the OpenMP5 backend (internally called
+`CPU_BT_OMP4` for now), one has to tell CMake explicitly about the OpenMP
```
`CPU_BT_OMP4` still true?

jkelling replied:
No, this got changed during renaming. The interjection in parenthesis is obsolete now, removed.

BenjaminW3 commented on README_OMP5.md:
```diff
@@ -0,0 +1,190 @@
+  -DALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLE=off \
+  -DALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE=off \
+  -DALPAKA_ACC_GPU_CUDA_ENABLE=off \
+  -DALPAKA_ACC_GPU_HIP_ENABLE=off \
```
It should not be necessary anymore to disable any backends on the command line now anymore. This is the default now. 
`cmake -DOpenMP_CXX_VERSION=5 -DALPAKA_ACC_ANY_BT_OMP5_ENABLE=ON` should be enough.

jkelling replied:
Also need to add `-DBUILD_TESTING=on -Dalpaka_BUILD_EXAMPLES=on` now (the README is currently for testers not users). Updated.

BenjaminW3 commented on .github/workflows/ci.yml:
```diff
@@ -64,7 +64,7 @@ env:
   ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE: ON
   ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE: ON
-  ALPAKA_ACC_CPU_BT_OMP4_ENABLE: ON
+  ALPAKA_ACC_ANY_BT_OMP5_ENABLE: OFF
```
As you are defaulting to `OFF` here, you can remove all the other mentions of `ALPAKA_ACC_ANY_BT_OMP5_ENABLE: OFF` below.

jkelling replied:
Done.

jkelling committed:
TEST_UNIT_KERNEL_KERNEL_STD_FUNCTION: Add clang unused macro WA

jkelling committed:
omp5: use MemsetKernel from mem/buf/SetKernel.hpp

move MemsetKernel to mem/buf/SetKernel.hpp was rebased before creation
of original omp4 version
... for reuse by OpenACC

jkelling committed:
alpakaCommon.cmake: Fix Omp5 checks

jkelling committed:
QueueOmp5*Blocking: Rm superflous #includes

jkelling committed:
update README_OMP5: new cmake defaults

jkelling committed:
install_clang.sh: install omp target packes based on OMP5_ENABLE

.. not based on clang version

jkelling commented:
> The selection which jobs are required is part of repository settings. They are mentioned there by name.
> This is somehow fragile but it is what github offers at the moment. When we want to rename a job, we have to change the settings as well. I can do this, once this PR is ready or has been merged.

Ok. My rationale about changing this name is, that this is the only job with omp5 enabled, and will be one of only a few for a while. If for a future PR clang-10 fails while others succeed, it may be good to see, that it includes the omp5 backend.

jkelling committed:
Omp5: re-enable multi-object test for GCC 11: linker bug fxed

jkelling committed:
test: queue: Add QueueGenericThreads*Blocking<DevOmp4>

jkelling commented:
> Could we use this PR as a general discussion board but split off some uncritical changes into independent PRs that can easily be merged?
> The first change I would like to get merged independently is the change you did to `EventHostManualTriggerCpuImpl`. I may find more things later.

https://github.com/alpaka-group/alpaka/pull/1133

BenjaminW3 commented on cmake/alpakaCommon.cmake:
```diff
@@ -234,20 +235,22 @@ endif()
 
-        # Clang versions starting from 3.9 support OpenMP 4.0 only when given the corresponding flag
-        if(ALPAKA_ACC_CPU_BT_OMP4_ENABLE)
+        # Clang versions starting from 3.9 support OpenMP 5.0 only when given the corresponding flag
```
```suggestion
        # Clang versions starting from 3.9 support OpenMP 4.0 and higher only when given the corresponding flag
```

BenjaminW3 commented on cmake/alpakaCommon.cmake:
```diff
@@ -234,20 +235,22 @@ endif()
-        if(ALPAKA_ACC_CPU_BT_OMP4_ENABLE)
+        # Clang versions starting from 3.9 support OpenMP 5.0 only when given the corresponding flag
+        if(ALPAKA_ACC_ANY_BT_OMP5_ENABLE)
             target_link_options(alpaka INTERFACE $<$<CXX_COMPILER_ID:AppleClang,Clang>:-fopenmp-version=40>)
```
Do we need to use `-fopenmp-version=50`? But which clang versions support this?
As far as I understood your PR, we do not use OpenMP 5.0 features, so we do not need to increase the version yet.

jkelling replied:
I know that clang 11 claims to support OpenMP 5.0 (it is incomplete), I did not check where exactly clang 10 is at.

This PR is not explicitly using OpenMP 5.0, but it relies on the compiler being lenient with some of OpenMP < 5.0's restrictions, chiefly the `static constexpr` member thing: https://github.com/alpaka-group/alpaka/pull/1118 .

jkelling replied:
The Math test uses `static constexpr` members, which is why it is disabled for GCC.

BenjaminW3 commented on include/alpaka/acc/AccOmp5.hpp:
```diff
@@ -58,36 +55,36 @@ namespace alpaka
+            public block::shared::st::BlockSharedMemStOmp5,
             public block::sync::BlockSyncBarrierOmp,
-            public intrinsic::IntrinsicCpu,
+            // cannot determine which intrinsics are save to use (depends on target), using fallback
```
```suggestion
            // cannot determine which intrinsics are safe to use (depends on target), using fallback
```

BenjaminW3 commented on include/alpaka/acc/AccOmp5.hpp:
```diff
@@ -147,36 +139,45 @@ namespace alpaka
+                    alpaka::ignore_unused(dev);
+
+#if defined(ALPAKA_OFFLOAD_MAX_BLOCK_SIZE) && ALPAKA_OFFLOAD_MAX_BLOCK_SIZE>0
+                    auto const blockThreadCount = ALPAKA_OFFLOAD_MAX_BLOCK_SIZE;
```
Is `ALPAKA_OFFLOAD_MAX_BLOCK_SIZE` meant to only reduce the `blockThreadCount` when there are too many available?
Should we then do `std::min(::omp_get_max_threads(), ALPAKA_OFFLOAD_MAX_BLOCK_SIZE)`?

jkelling replied:
Yes, this is sensible.

BenjaminW3 commented on include/alpaka/acc/AccOmp5.hpp:
```diff
@@ -147,36 +139,45 @@ namespace alpaka
 #else
-                    auto const blockThreadCountMax(alpaka::core::clipCast<TIdx>(::omp_get_max_threads()));
+                    auto const blockThreadCountMax(alpaka::core::clipCast<TIdx>(blockThreadCount));
+                    auto const gridBlockCountMax(alpaka::core::clipCast<TIdx>(::omp_get_max_threads())); //! \todo fix max block size for target
```
What do you want to do with the TODO here?

jkelling replied:
This is basically a todo for a later OpenMP version: On CPU we would want `blockThreadCountMax * gridBlockCountMax` to equal the number of CPU threads available, so we would write:
```c++
auto const gridBlockCountMax(alpaka::core::clipCast<TIdx>(::omp_get_max_threads() / blockThreadCountMax));
```
However, If we did this, we would always run with one block (if `ALPAKA_OFFLOAD_MAX_BLOCK_SIZE` does not say otherwise) even on GPU, because `omp_get_max_threads()` appears to return the number of threads per block/team when the target is a GPU (at least XL does that, which is the only compiler with which I was actually able to run any of this on a GPU).

What we really want to do is `gridBlockCountMax =  omp_get_max_teams()`, but this function does not exist.

jkelling replied:
The XL behavior may not be conforming to OpenMP 5.0, but I could not test with any other compiler, so any fix for this would at this point be theoretical. The way it is currently handled does not break too much. It can be fixed once it is clear how compilers handle this.

BenjaminW3 replied:
Could you add parts of your answer to an extended comment?

jkelling replied:
https://github.com/alpaka-group/alpaka/pull/1126/commits/fbf0df054a673698127df84c73342514bbcb6290

BenjaminW3 commented on include/alpaka/core/Omp5.hpp:
```diff
@@ -0,0 +1,72 @@
+
+#include <omp.h>
+
+#include <array>
```
where do we need `<array>` here?

jkelling replied:
No.

BenjaminW3 commented on include/alpaka/core/Omp5.hpp:
```diff
@@ -0,0 +1,72 @@
+#include <alpaka/idx/Traits.hpp>
+#include <alpaka/vec/Vec.hpp>
+#include <alpaka/meta/IntegerSequence.hpp>
+#include <alpaka/meta/Metafunctions.hpp>
```
This block of includes also seems unused.

BenjaminW3 commented on include/alpaka/dev/DevOmp5.hpp:
```diff
@@ -0,0 +1,365 @@
+        namespace traits
+        {
+            //#############################################################################
+            //! The CUDA RT device name get trait specialization.
```
There are multiple `CUDA RT` in the comments in this file.

BenjaminW3 commented on include/alpaka/dev/DevOmp5.hpp:
```diff
@@ -0,0 +1,365 @@
+                {
+                    alpaka::ignore_unused(dev); //! \todo query device
+                    std::size_t freeInternal((6ull<<30));
+                    // std::size_t totalInternal(0u);
```
Any ideas here and above?

jkelling replied:
The OpenMP standard does not offer a way to query this. AFAIK, there is also no way to figure what type of device the target is, so we could check if we are targeting the CPU, then check the host memory, or if the we are targeting the GPU and then ask the CUDA/HIP runtime (which would be problematinc any because it would become a dependency). Thought of a workaround like this was the reason for the `\todo`. I think this can just not be supported. What is the alpaka policy to report, that this number is not available?

BenjaminW3 replied:
Any non-zero value might be too much on a given hardware.
Only pessimistic answers can be valid answers. As we do not have any hint about the real sizes, zero seems to be the only safe return value.

jkelling replied:
Done.

jkelling replied:
@psychocoderHPC : Reporting zero here breaks PIConGPU via pmacc, which apparently checks this and then fails to allocate memory because there are only 0B left. How do you suggest to resolve this? Would you change this to be more optimistic (basically `INT64_MAX`) or should this be handled in pmacc?

jkelling replied:
@BenjaminW3 Maybe there should be a trait added to check if the value returned by this one can be trusted on the given backend or should be ignored.

BenjaminW3 replied:
Or return an optional

jkelling replied:
Using optional would be a breaking change. With this:
```c++
template<class TDev>
ALPAKA_FN_HOST static auto getMemBytes(TDev const & dev)
 -> boost::optional<std::size_t>
{...}
```
All uses would need to change to:
```c++
auto memBytes = getMemBytes(dev);
if(memBytes)
{
 if(*memBytes < howMuchWeNeed)
// ...
}
else
{
// hope there is enough
}
```

BenjaminW3 replied:
But better explicitly break the interface to make users aware than hiding the still breaking change by reporting UINT_MAX or similar unknown values.
Lets wait for an answer from @psychocoderHPC. I would still prefer returning zero.

jkelling replied:
I am with you on the zero vs. UINT_MAX issue.

Adding another trait would however be less breaking than returning an optional: Using optional would require all code using any backend to change. Adding another trait, would allow legacy code work unchanged with all backends other than this one. Only code that wants to support this new backend would have to add a check with the new trait ( `deviceKnowsAvailableMem` or something ), which with time would become all code, but it could happen more gradually.

psychocoderHPC replied:
Is it possible to start a kernel to figure out which type of device is used? Than we could use the return value of this kernel to select they way how to calculate the available memory.

jkelling replied:
I do not see any standard way to get the type of device being used. We could add a cmake option to select the target which could then define a macro and set the appropriate compiler flags. Something like this should become part of the OpenMP cmake module at some point. This would, however, not work when the stand-alone headers are used.

Even if we could determine what type of device is being used, there would still be no guarantee, that we can determine a value, because we need code for each device type, so we would have nothing in a new environment.

We need a way to tell the program, that we cannot get this information as a fallback.

jkelling replied:
Also: If this would be a runtime decision, all strategies for possible targets need to be compiled. While it is possible to leave out runtimes which are not available at compile time, this would make CUDA a dependency if you want to compile for OpenMP 5 targeting nvptx,for example. Usually this dependency would not be explicit and it might sometimes be hidden away in the compiler.

jkelling replied:
I had a discussion with @psychocoderHPC 
* The PIConGPU/pmacc issue is not central and can be worked around very easily
* He brought up the option of raising an exception in the specialization, which would only break application if they use this backend and the root cause would be easy to track down. Also these functions are not pefrormance critical.
* He would be OK with just returning zero. I will add a commit to this PR documenting the possibility of this trait returning 0, which then means, that the real value is unknown. I check this issue for the OpenACC backend: There it is possible to get these values (works with NHVPC 20.7+GPU). However GCC just returns 0, which maybe because of a lack of support or disabled offloading, and a lack of support in the runtime for the host. This means, 0 return values can also occur for this trait when support is technically there.

jkelling replied:
Added https://github.com/alpaka-group/alpaka/pull/1126/commits/ab9b8dbb094a67703bc627076f1652a6da12681e

I suggest to move forward with this state and optionally create an issue in case we want to consider more invasive handling (optional, exception, ...).

BenjaminW3 commented on include/alpaka/dev/DevOmp5.hpp:
```diff
@@ -0,0 +1,365 @@
+                    ALPAKA_DEBUG_FULL_LOG_SCOPE;
+
+                    generic::currentThreadWaitForDevice(dev);
+// #pragma omp taskwait
```
?

jkelling replied:
Either this is left over from the old OpenMP4, or I was thinking about implementing an OpenMP tesking-based queue. Removed.

BenjaminW3 commented on include/alpaka/idx/bt/IdxBtOmp5BuiltIn.hpp:
```diff
@@ -0,0 +1,145 @@
+                    alpaka::ignore_unused(idx);
+                    return vec::Vec<dim::DimInt<1u>, TIdx>(static_cast<TIdx>(omp_get_thread_num()));
+                }
+            };
```
Could you add this 1D specialization to `alpaka::idx::bt::IdxBtOmp` and get rif of `IdxBtOmp5BuiltIn` completely?
This could also be a separate PR.

jkelling replied:
https://github.com/alpaka-group/alpaka/pull/1140

BenjaminW3 commented on include/alpaka/mem/view/Traits.hpp:
```diff
@@ -115,8 +115,9 @@ namespace alpaka
-                                extent::getExtent<dim::Dim<TView>::value - 1u>(view)
-                                * sizeof(elem::Elem<TView>);
+                                static_cast<idx::Idx<TView>>(
+                                    static_cast<unsigned int>(extent::getExtent<static_cast<unsigned int>(dim::Dim<TView>::value) - 1u>(view))
```
What exactly made it necessary to cast the non-type-template parameters here?

jkelling replied:
I got warnings there and it was either cast `1u` to `Idx` or cast the rest to `unsigned int`, I went with the latter. I think there are cases where `idx::Idx<TView>` is signed, maybe creating warnings because of the unsigned value returned by `sizeof`.

BenjaminW3 replied:
The `getExtent<static_cast<unsigned int>` seems to be valid.
The outer `static_cast<idx::Idx<TView>>` makes it explicit what the `return` will do anyways.
However, the `static_cast<unsigned int>(extent::getExtent<` is dangerous because we should not change / clip the values returned by `getExtent`.

jkelling replied:
Maybe test causing this issue were changed in the meantime, if I revert this hunk, I can now build without warnings.

Should I remove this change or just reduce it to the innermost cast to `unsigned int`?

BenjaminW3 replied:
I would completely remove it

BenjaminW3 commented on include/alpaka/mem/view/ViewPlainPtr.hpp:
```diff
@@ -349,6 +350,39 @@ namespace alpaka
+                        TExtent const & extent)
+#ifdef BOOST_NO_CXX14_RETURN_TYPE_DEDUCTION
+                    -> alpaka::mem::view::ViewPlainPtr<dev::DevOmp5, TElem, alpaka::dim::Dim<TExtent>, alpaka::idx::Idx<TExtent>>
+#endif
```
This `#ifdef` can be removed now.

BenjaminW3 commented on include/alpaka/pltf/PltfOmp5.hpp:
```diff
@@ -0,0 +1,105 @@
+                    std::size_t const devCount(static_cast<std::size_t>(::omp_get_num_devices()));
+                    int devIdxOmp5 = static_cast<int>(devIdx);
+                    if(devIdx >= devCount)
+                    { // devIdx param must be unsigned, take take this case to use the initial device
```
"take take"?
The other platforms throw an exception in the case of an invalid `devIdx`. We should do this here as well.

jkelling replied:
OpenMP target has target devices with `0 <= devIdxOmp5 < devCount`. In addition, there is an "initial device", which has a `devIdxOmp5` which is implementation defined, all compilers I saw have `devIdxOmp5 < 0`. The trait's parameter is size_t, so we cannot map this directly, so I decided to use `devIdx > devCount`. Apparently, they want change this in OpenMP 5.1 to mandate `initialDevice == devCount`.

BenjaminW3 replied:
`devIdx` and `devCount` are in general totally unrelated to `devIdxOmp5`. We could have an arbitrary mapping between those.
I would imagine that we always artificially add one to the value returned by `omp_get_num_devices()` to account for the default-device when giving the user the number of available devices. `devIdx == omp_get_num_devices()` would request this default device. All other indices would directly map to the `devIdxOmp5` and `devIdx > omp_get_num_devices()` would result in the exception.
We could also change this mapping to make `devIdx == 0` the default-device and offset all others by one.

jkelling replied:
Including the initial device in `devCount` would break performance of programs which map ranks to device in a wrap-around fashion or which try to use all devices reported by alpaka. The normal use case would be only use the actual target devices and not the host, i.e. `0 <= devIdx < devCount`. Only if there are not target devices available it makes sense include the host and thus have `devCount = 1` instead of `0`, merging OpenMP behavior with normal alpaka-backend behavior. There is no alpaka interface for using the initial device/host through the backend. Doing this, when there are actual target devices available, is purely an OpenMP target thing, thus only this case should deviate from normal alpaka behavior. If the user code is not aware which backend is being used, it should be able expect the same behavior from all backends, as long as it uses the backends properly through the interface.

We could limit this to using the initial device only if `devIdx == omp_get_num_devices()`, because this can be elegantly checked, and throw on other values.

This would of course still lead to the same unexpected behavior if user code were to iterate the devices like this:
```c++
try
{
 for(int a; ;++a)
 {
  getDevice(a);
  //...
 }
}
catch (out_of_range){}
```
but I think we do not need to cater to bad style.

BenjaminW3 replied:
> We could limit this to using the initial device only if devIdx == omp_get_num_devices(), because this can be elegantly checked, and throw on other values.

Sounds good

jkelling replied:
https://github.com/alpaka-group/alpaka/pull/1126/commits/2db34b82d4586a5ea853005aba532e9b5ddfccf2

In addition to the above condition, it also uses the initial device if `devCount == 0 && devIdx == 1`, that is when `devIdx` is equal to the return value of `getDevCount()`, which is this case differs from `omp_get_num_devices()`.

BenjaminW3 commented on include/alpaka/pltf/PltfOmp5.hpp:
```diff
@@ -0,0 +1,105 @@
+                    ALPAKA_DEBUG_FULL_LOG_SCOPE;
+
+                    std::size_t const devCount(static_cast<std::size_t>(::omp_get_num_devices()));
+                    int devIdxOmp5 = static_cast<int>(devIdx);
```
What is the mapping between alpaka `devIdx` and OpenMP? Is `devIdx == 0` always the host and `devIdx > 0` the offloading devices?
Can we document this?

jkelling replied:
https://github.com/alpaka-group/alpaka/commit/2db34b82d4586a5ea853005aba532e9b5ddfccf2#diff-0af16922c3afe4ac9df8b1a1c77c3229R84

Changed the docstring a bit.

BenjaminW3 commented on script/docker_ci.sh:
```diff
@@ -57,9 +65,13 @@ if [ ! -z "${ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE+x}" ]
+if [ ! -z "${ALPAKA_OFFLOAD_MAX_BLOCK_SIZE+x}" ]
 then
-    ALPAKA_DOCKER_ENV_LIST+=("--env" "ALPAKA_ACC_CPU_BT_OMP4_ENABLE=${ALPAKA_ACC_CPU_BT_OMP4_ENABLE}")
+    ALPAKA_DOCKER_ENV_LIST+=("--env" "ALPAKA_OFFLOAD_MAX_BLOCK_SIZE=${ALPAKA_OFFLOAD_MAX_BLOCK_SIZE}")
```
Is `ALPAKA_OFFLOAD_MAX_BLOCK_SIZE` limited to OpenMP for now?
Should we rename it to `ALPAKA_OMP_OFFLOAD_MAX_BLOCK_SIZE`?

jkelling replied:
It is also used by OpenAcc.

BenjaminW3 commented on include/alpaka/dev/DevOmp5.hpp:
```diff
@@ -0,0 +1,365 @@
+                private:
+                    std::mutex mutable m_Mutex;
+                    std::vector<std::weak_ptr<queue::IGenericThreadsQueue<DevOmp5>>> mutable m_queues;
+                    int m_iDevice = 0;
```
Can we make this `const` and remove the `=0`? It is always initialized in the constructor.

jkelling committed:
omp5: Fix version comments in alpakaCommon.cmake

jkelling committed:
Fix comment spelling in AccOmp5.hpp

Co-authored-by: Benjamin Worpitz <benjaminworpitz@gmail.com>

jkelling commented on include/alpaka/dev/DevOmp5.hpp:
```diff
@@ -0,0 +1,365 @@
+        {
+            friend struct pltf::traits::GetDevByIdx<pltf::PltfOmp5>;
+
+        protected:
+            //-----------------------------------------------------------------------------
+            DevOmp5(int iDevice) :
```
Just noticed, that this ctor is protected, not private. I do not currently see a use case for deriving from this class and using this ctor in the derived class. So, I will remove the `protected:` from this PR.

jkelling committed:
Omp5: Code cleanup

jkelling committed:
AccOmp5: Use IdxBtOmp

jkelling committed:
AccOmp5: document choice of gridBlockCountMax

jkelling committed:
getDevByIdx<PltfOmp5>: restrict devIdx contitions...

for initial_device, throw for devIdx > devCount

jkelling committed:
CI: Add omp5 gcc10 analysis build

Rename -omp5 jobs to _omp5

BenjaminW3 commented on .github/workflows/ci.yml:
```diff
@@ -81,85 +81,88 @@ jobs:
       matrix:
         include:
         ### Analysis builds
+        - name: linux_gcc-10_debug_analysis-omp5
```
Could you please use an underscore here and in the renamed clang job?
```suggestion
        - name: linux_gcc-10_debug_analysis_omp5
```

jkelling replied:
Done,.

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Set.hpp:
```diff
@@ -0,0 +1,116 @@
+                        byteExtent[TDim::value-1] *= static_cast<Idx>(sizeof(elem::Elem<TView>));
+                        constexpr auto lastDim = TDim::value - 1;
+
+                        if(pitch[0] <= 0)
```
Could you please add braces around the if block?

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Set.hpp:
```diff
@@ -0,0 +1,116 @@
+                                    pitch
+                                ); // NOP if size is zero
+
+                        std::cout << "Set TDim=" << TDim::value << " pitch=" << pitch << " byteExtent=" << byteExtent << std::endl;
```
Could you please guard this log line with `#if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL`?

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Set.hpp:
```diff
@@ -0,0 +1,116 @@
+                        byteExtent[TDim::value-1] *= static_cast<Idx>(sizeof(elem::Elem<TView>));
+                        constexpr auto lastDim = TDim::value - 1;
+
+                        if(pitch[0] <= 0)
```
What exactly is this pitch check doing here?
Shouldn't we be checking the `byteExtent.prod()`?
Furthermore, negative pitches are a valid thing. Some image formats describe their buffers with a pointer to the back and pitch being negative.

jkelling replied:
I noticed, that the existing pitch calculations always lead to `pitch[0]` containing `byteExtent.prod()`, so we would not need to calculate it here.

I was not aware of the possibility of negative pitches, could we check `pitch[0] != 0` instead?

BenjaminW3 replied:
yes, should be the correct thing to do.

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Set.hpp:
```diff
@@ -0,0 +1,116 @@
+                        auto pitch = view::getPitchBytesVec(view);
+                        auto byteExtent = extent::getExtentVec(extent);
+                        byteExtent[TDim::value-1] *= static_cast<Idx>(sizeof(elem::Elem<TView>));
+                        constexpr auto lastDim = TDim::value - 1;
```
Could you switch those two lines and reuse `lastDim` in the previous line?

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Set.hpp:
```diff
@@ -0,0 +1,116 @@
+                                    workdiv::WorkDivMembers<TDim, Idx>(
+                                        vec::Vec<TDim, Idx>::ones(),
+                                        vec::Vec<TDim, Idx>::ones(),
+                                        vec::Vec<TDim, Idx>::ones()),
```
Isnt't this the "nothing to do"-case? Shouldn't this be zeros?

jkelling replied:
A kernel with a zero workdiv currently leads to a divide-by-zero exception. I did not take zeros to be a valid workdiv.

As it is now, the kernel will run, see that the extent is zero and terminate. With zeros, the `TaskKernelOmp5::operator()` needs to check this and return without executing.

jkelling replied:
I pushed this change: https://github.com/alpaka-group/alpaka/pull/1126/commits/5626aae8e287102b91a52daef51dbe86212eecb0

jkelling replied:
Amended: https://github.com/alpaka-group/alpaka/pull/1126/commits/50666f6941193635a1d1f4f226de3ec5261b7655
(Make comment doc string)

jkelling replied:
@psychocoderHPC , @BenjaminW3 : Do you agree with this behavior of `TaskKernelOmp5::operator()`, i.e. return without doing anything and not raising an error if the grid is empty?

BenjaminW3 replied:
I am OK with this

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Copy.hpp:
```diff
@@ -0,0 +1,545 @@
+                                // std::cout << "copy " << TDim::value << "d\textent=" << m_extent
+                                //     << "\tdstExtentFull=" << dstExtentFull << " (p " << m_dstPitchBytes
+                                //     << " )\tsrcExtentFull=" << srcExtentFull << " (p " << m_srcPitchBytes
+                                //     << " )\telementSize=" << elementSize << std::endl;
```
remove those comments?

BenjaminW3 commented on include/alpaka/mem/buf/omp5/Copy.hpp:
```diff
@@ -0,0 +1,545 @@
+                                        MyGetPitch>::vecFromDimTrait(viewDst)),
+                                m_srcPitchBytes(VecFromDimTrait<
+                                        TDim, size_t, TViewSrc,
+                                        MyGetPitch>::vecFromDimTrait(viewSrc)),
```
What is the difference to the `mem::view::getPitchBytesVec(viewSrc)`, that is used in ` TaskCopyCpuBase` (and `getExtentVec` for above)?

jkelling replied:
This version is casting to a `Vec<size_t>`. The class stores the extents with this type because they are directly used as arguments to the omp runtime functions, which expect arrays of `size_t`. An alternative to this may be to add a cast operator to `Vec<..>`.

BenjaminW3 replied:
You can find it here: https://github.com/alpaka-group/alpaka/blob/d1f7047ae2116cc6a83674ce352adabbb263ad04/test/unit/vec/src/VecTest.cpp#L91-L113

jkelling replied:
Thanks; I put it in.

BenjaminW3 commented on include/alpaka/mem/buf/BufOmp5.hpp:
```diff
@@ -0,0 +1,713 @@
+                {}
+
+                    BufOmp5(const BufOmp5&) = default;
+                    BufOmp5(BufOmp5 &&) = default;
```
indentation seems to be wrong

jkelling replied:
Habit (indenting ctors and dtors), sorry. Fixed.

BenjaminW3 commented on include/alpaka/mem/buf/BufOmp5.hpp:
```diff
@@ -0,0 +1,713 @@
+                            BufOmp5Impl(BufOmp5Impl &&) = default;
+                        BufOmp5Impl& operator=(const BufOmp5Impl&) = delete;
+                        BufOmp5Impl& operator=(BufOmp5Impl&&) = default;
+                            ~BufOmp5Impl()
```
indentation wrong here and above as well

BenjaminW3 commented on include/alpaka/kernel/TaskKernelOmp5.hpp:
```diff
@@ -0,0 +1,386 @@
+                TIdx const blockThreadCount(blockThreadExtent.prod());
+
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+                if(maxOmpThreadCount < blockThreadExtent.prod())
```
braces here please as well.

BenjaminW3 commented on test/unit/atomic/CMakeLists.txt:
```diff
@@ -20,5 +20,6 @@ target_link_libraries(
     PRIVATE common)
 
 set_target_properties(${_TARGET_NAME} PROPERTIES FOLDER "test/unit")
+target_compile_definitions(${_TARGET_NAME} PRIVATE "-DTEST_UNIT_ATOMIC")
```
Should we do this confitionally only when `ALPAKA_ACC_ANY_BT_OMP5_ENABLE` is set? This would also automatically document why this is necessary.

jkelling replied:
Doing this would require this line to change again in https://github.com/alpaka-group/alpaka/pull/1127 , because they need to be excluded there.

I would rather add a line like this to all test, so each test can be identified by the code in accelerator selection code [test/common/include/alpaka/test/acc/TestAccs.hpp](https://github.com/alpaka-group/alpaka/pull/1126/files#diff-c93c0e3f72fa0a6945dfff2600feafd4). Other new backends, like [SYCL](https://github.com/alpaka-group/alpaka/pull/789) may also want to exclude some test at some stages of development. This could be a saparate PR, then.

BenjaminW3 commented on test/integ/separableCompilation/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@
 #
 
-if((NOT ALPAKA_ACC_GPU_CUDA_ENABLE AND NOT ALPAKA_ACC_GPU_HIP_ENABLE) OR
+if((NOT ALPAKA_ACC_GPU_CUDA_ENABLE AND NOT ALPAKA_ACC_GPU_HIP_ENABLE AND NOT ALPAKA_ACC_ANY_BT_OMP5_ENABLE ) OR
```
separable compilation is not supported? Is this also the known linker error that will be fixed with gcc-11?

jkelling replied:
The issue here is that functions which are not defined inline have to be enclosed in
```c++
#pragma omp declare target
...
#pragma omp end declare target
```
so target code is generated and it can be called  in target regions. This cannot be worked into the `ALPAKA_FN_HOST_DEVICE` macro. Note, even if there would only be one pragma in front required, like in OpenACC, we could still not map this to the macro, because this would break template declarations:
```c++
template<class T>
ALPAKA_FN_DEVICE void f(int);
```
would expand to
```c++
template<class T>
_Pragma("acc routine seq") void f(int);
```
but you cannot place the pragma between the template and the function part.

BenjaminW3 replied:
Is this missing support for separable compilation documented somewhere?

jkelling replied:
No, I only disabled the test. Which would be the place to document this?

BenjaminW3 replied:
It should probably be part of `README_OMP5.md`

jkelling replied:
Added: https://github.com/alpaka-group/alpaka/pull/1126/commits/9e78aa35b9138cc5a096d5e597833554c9662623
(also some updates on compiler status)

BenjaminW3 commented on test/unit/kernel/src/KernelStdFunction.cpp:
```diff
@@ -7,6 +7,11 @@
+#define TEST_UNIT_KERNEL_KERNEL_STD_FUNCTION
+// clang thinks a macro is unused when only used as the second operand of an &&
+// where the first operand evaluates to false, so we use it here:
+TEST_UNIT_KERNEL_KERNEL_STD_FUNCTION
```
Why is this clang workaround not necessary for the other tests?
Should we remove one `KERNEL` from the name?

jkelling replied:
1. Maybe because there the macros are defined via `-D` instead of in the code and are thus maybe not triggering warnings.
2. This one only disables a part of what would be called `TEST_UNIT_KERNEL`, namely the `KERNEL_STD_FUNCTION` part. This is also why this is defined in the source file and not via `target_compile_definitions`.

jkelling commented:
Somehow adding that one CI job "linux_gcc-10_debug_analysis_omp5" made the number of check go up from 101 to 107. Did this create some other jobs implicitly due to the "matrix" mode?

jkelling committed:
TaskKernelOmp5: Add treat empty grid as NOP

set<BufOmp5>: return empty-grid kernel if range empty

BenjaminW3 commented:
Did you rebase which might have added some new builds to your branch?

jkelling commented:
> Did you rebase which might have added some new builds to your branch?

A diff to develop shows many changes, but mostly removing `ALPAKA_ACC_CPU_BT_OMP4_ENABLE=OFF`. "linux_gcc-10_debug_analysis_omp5" is the only one added.

jkelling commented:
... or were jobs recently added to develop? I did rebase onto the current state of the develop branch.

BenjaminW3 commented:
I think some Windows+CUDA 11 builds have been added

jkelling committed:
dev::getMemBytes: Add doc about valid 0 return

BenjaminW3 commented:
I do not think that more than 100 commits bring any value. Could you please reduce the number of commits (maybe to one)?

jkelling commented:
I can squash after all issues are resolved.

jkelling committed:
Update README_OMP5

BenjaminW3 commented:
@psychocoderHPC Could you please also review this PR?

psychocoderHPC commented on include/alpaka/acc/AccOmp5.hpp:
```diff
@@ -58,36 +55,36 @@ namespace alpaka
             public idx::bt::IdxBtOmp<TDim, TIdx>,
             public atomic::AtomicHierarchy<
-                atomic::AtomicStdLibLock<16>,   // grid atomics
+                atomic::AtomicOmpBuiltIn,   // grid atomics
```
Is OpenMP guarantee that atomic is atomic between grids?
If not than it would not be allowed to use atomics in two parallel running kernel where the same memory is accessed. We need to at at least a warning about unexpected behaviour into the documentation.

jkelling replied:
It depends:
OpenMP 5 2.17.7 . Binding
```
If the size of x is 8, 16, 32, or 64 bits and x is aligned to a multiple of its size, the binding thread set
for the atomic region is all threads on the device. Otherwise, the binding thread set for the
atomic region is all threads in the contention group. atomic regions enforce exclusive access
with respect to other atomic regions that access the same storage location x among all threads in
the binding thread set without regard to the teams to which the threads belong.
```

sbastrakov replied:
Did I understand the quoted fragment right that in case the size and alignment conditions are met, atomicity is guaranteed for all threads on a device (so between grids in alpaka speech), while if they are not met, only between threads of an openmp team (not sure what it corresponds to in alpaka for this backend)?

jkelling replied:
That is how I understood it.

psychocoderHPC reviewed (changes requested):
There is only one open question about atomic functions between grids.

psychocoderHPC commented:
Welcome in the jungle of alpaka backends
